crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_structure.rs

```diff
@@ -125,4 +165,109 @@ mod tests {
             }
         }
     }
+    #[test]
+    fn we_can_correctly_identify_index_existence() {
+        assert!(index_exists(0));
+        assert!(!index_exists(no_index()));
+        assert!(index_exists(no_index() - 1));
+    }
+    #[test]
+    fn we_can_find_the_full_width_of_row() {
+        let width_of_row = full_width_of_row(0);
+        assert_eq!(width_of_row, 1);
+
+        let width_of_row = full_width_of_row(1);
+        assert_eq!(width_of_row, 2);
+
+        let width_of_row = full_width_of_row(4);
+        assert_eq!(width_of_row, 4);
+
+        let width_of_row = full_width_of_row(7);
+        assert_eq!(width_of_row, 8);
+
+        let width_of_row = full_width_of_row(11);
+        assert_eq!(width_of_row, 8);
+
+        let width_of_row = full_width_of_row(12);
+        assert_eq!(width_of_row, 16);
+
+        let width_of_row = full_width_of_row(13);
+        assert_eq!(width_of_row, 16);
+    }
+    #[test]
+    fn we_can_produce_the_correct_matrix_size() {
+        let (width, height) = matrix_size(2, 0);
+        assert_eq!(width, 2);
+        assert_eq!(height, 2);
+
+        let (width, height) = matrix_size(3, 0);
+        assert_eq!(width, 2);
+        assert_eq!(height, 3);
+
+        let (width, height) = matrix_size(4, 0);
+        assert_eq!(width, 2);
+        assert_eq!(height, 3);
+
+        let (width, height) = matrix_size(16, 0);
+        assert_eq!(width, 4);
+        assert_eq!(height, 6);
+
+        let (width, height) = matrix_size(17, 0);
+        assert_eq!(width, 8);
+        assert_eq!(height, 7);
+
+        let (width, height) = matrix_size(64, 0);
+        assert_eq!(width, 8);
+        assert_eq!(height, 12);
+
+        let (width, height) = matrix_size(71, 0);
+        assert_eq!(width, 16);
+        assert_eq!(height, 13);
+    }
```

Comment:
Can you do this exhaustively? I'd like to see it at least 0..16 with some bigger numbers like you have here as well.

Reply:
I listed test cases for `0..16` explicitly with a few extra cases. I could only think of recreating the algorithm in the test case to do it exhaustively for a larger number, which didn't seem appropriate for a test case. Should I spend some more time thinking of an exhaustive `0..N` test?

Reply:
No, don't worry about an exhaustive test. I just meant explicitly.